### PR TITLE
feat(tests): Shamir backup walkthrough function and test

### PR DESCRIPTION
**Pull request by grdddj**
Supported Shamir backup functionality for Suite tests
- connected with [trezor-user-env PV](https://github.com/trezor/trezor-user-env/pull/89)

NOTE: When running this "2 out of 3" Shamir, at the end the `websocket_timeout` was encountered

### CODE REVIEW

**Comment by tsusanka** [comment]
> NOTE: When running this "2 out of 3" Shamir, at the end the websocket_timeout was encountered

We should investigate this thoroughly as this might be the cause for the occasional flakyness maybe?

**Comment by tsusanka** [comment]
Other than that LGTM, but I did not test and I think @tomasklim or @mroz22 should have a look.

**Comment by grdddj** [comment]
> We should investigate this thoroughly as this might be the cause for the occasional flakyness maybe?

When doing `2 out of 2` backup, the error is not thrown, so it will be most probably due to prolonged action on the emulator (it takes some time to click everything, as we wait 0.5 seconds after each action on emulator)

**Comment by tsusanka** [comment]
>  (it takes some time to click everything, as we wait 0.5 seconds after each action on emulator)

In Cypress right? I mean we wait 0.5 seconds in cypress?

**Comment by tsusanka** [comment]
The ideal state would be to somehow connect it to some event in emulator and wait for that event. Can't cypress wait until the response from trezor-user-env is sent? Or do we send the resposne right away even if the action is not finished yet? :thinking:

**Comment by grdddj** [comment]
> In Cypress right? I mean we wait 0.5 seconds in cypress?

I am not aware of Cypress waits, but we definitely wait `0.501 s` after each action in [trezor-user-env emulator.py](https://github.com/trezor/trezor-user-env/blob/b7ddfa9616a92fce2d3d38c12fed9d9ed798a61d/src/emulator.py#L24-L28) for reasons described there (not by me)

Therefore the process of multiple shares backup takes long time and seems to cause websocket timeout.

However, this timeout should be configurable, so we can increase it.

**Comment by grdddj** [comment]
> The ideal state would be to somehow connect it to some event in emulator and wait for that event. Can't cypress wait until the response from trezor-user-env is sent? Or do we send the resposne right away even if the action is not finished yet? thinking

The former is already the case - controller receives request `perform shamir backup simulation` and it starts doing so, and after it finishes (let's say after 15 seconds), only then it sends the response.

Because the controller (websocket) is "frozen" during the emulator action, it looks as inactive and the connection with the client seems to timeout after some time (15 seconds is a long time for a computer).

We could theoretically incorporate some multithreading into the controller, but it could mean other complications.

**Comment by tsusanka** [comment]
Hmm. I think this will need some bigger overhaul.
- What is the timeout on the cypress side? I think it is indeed 15s.
> I am not aware of Cypress waits, but we definitely wait 0.501 s after each action in trezor-user-env emulator.py
- This is weird. And I am not sure it is needed. @mroz22 did you write or do you remember it? `git blame` does not help here, it was copied from Suite.
- If some action takes longer time in trezor-user-env we should send some keep-alive responses. Let's say after 0.5s or similar. But that is definitely something we need to discuss with @mroz22 or someone from the Suite team. Maybe I should schedule a meeting?

**Comment by grdddj** [comment]
> Maybe I should schedule a meeting?

I would be glad for that.


What we can do as an experiment - to decrease (or even remove) this `sleep` interval in our `incoming` branch and run all the Suite tests in CI.

**Comment by tsusanka** [comment]
> What we can do as an experiment - to decrease (or even remove) this sleep interval in our incoming branch and run all the Suite tests in CI.

Let's try that.

**Comment by grdddj** [comment]
> Let's try that.

After changing that to `0.251 s` we have some failing tests - https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/1447074651 - but hard to say what is the rootcause of that (it could be caused by misbehaving emulator or misbehaving Suite)

**Comment by grdddj** [comment]
Changing that to `0.101 s` caused issues with `readAndConfirmMnemonicEmu`. However, the "infamous" always failing backup test is running fine:
- [backup success](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/1447234589)
- [mnemonic and other failures](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/1447234584)

**Comment by grdddj** [comment]
Some observations from local testing of `onboarding/t2-create-wallet.test.ts` (which uses the above-mentioned `readAndConfirmMnemonicEmu`):
- the first test with `2.1.4` emulator version is failing when the SLEEP is `0.25 s` or less, the second one, `2.3.4 `emulator is fine even in that case (so could be that older emulators need this SLEEP and newer do not)
- in the first test, even after successfully doing the backup, there was an error message on the screen which caused the test to fail:
image

**Comment by tsusanka** [comment]
I would strongly consider removing such old firmwares from Suite e2e. `2.1.4` is two years old, which is exactly what we will stop supporting soon. I would test either on master or few months old firmware first and then we could introduce some "old firmwares test" that would test this, but most of the tests should be against new firmwares. There were many debuglink improvements and who knows what is happening in 2.1.4..

**Comment by grdddj** [comment]
Commit cf8fa05:
- as per [this discussion](https://github.com/trezor/trezor-suite/issues/4069#issuecomment-885640113_) increasing websocket timeout to 60 seconds, which makes the Shamir backup test succeed

File: packages/integration-tests/projects/suite-web/plugins/websocket-client.js
Comment (tsusanka) [inline packages/integration-tests/projects/suite-web/plugins/websocket-client.js]: Is this used only for trezor-user-env? Maybe we could add a comment. Something in the sense that the timeout is quite high because tenv is synchronous and some commands take some time.

**Review by tsusanka — APPROVED** [review]
One comment. Feel free to force-push then.

File: packages/integration-tests/projects/suite-web/plugins/websocket-client.js
Comment (tsusanka) [reply packages/integration-tests/projects/suite-web/plugins/websocket-client.js]: Btw how long does the Shamir command take on your computer? Let's say 3 of 5 Shamir.

Comment (grdddj) [reply packages/integration-tests/projects/suite-web/plugins/websocket-client.js]: Shamir 3 out of 5 (the 3 really does not play a role there in wallet creation process) took about 30 seconds.

Comment (tsusanka) [reply packages/integration-tests/projects/suite-web/plugins/websocket-client.js]: Ok let's see.

**Comment by tsusanka** [comment]
Last thing: the latest commit message now does not correspond to the actual timeout value :).

**Comment by grdddj** [comment]
> Last thing: the latest commit message now does not correspond to the actual timeout value :).

Right, let me change the commit message

**Review by tsusanka — APPROVED** [review]

**Pull request merged**
